test/hotspot/jtreg/compiler/intrinsics/string/TestHasNegatives.java

```diff
@@ -26,81 +26,99 @@
-            for (int i = 0; i < neg; ++i) {
-                idx = off + (len - 1) - div * i;
-                tBa[idx] = (byte) (0x80 | tBa[idx]);
+            for (int i = 0; i < neg; i++) {
+                int idx = off + RANDOM.nextInt(len);
+                bytes[idx] = (byte) (0x80 | bytes[idx]);
             }
         }
         // insert "canary" negative values after array segment
-        for (int i = len + off; i < tBa.length; ++i) {
-            tBa[i] = (byte) (((i + 15) & 0x7F) | 0x80);
+        for (int i = len + off; i < bytes.length; ++i) {
+            bytes[i] = (byte) (((i + 15) & 0x7F) | 0x80);
         }
     }
 
-    /** Sizes of array segments to test. */
-    private static int sizes[] = { 1, 2, 3, 4, 5, 6, 7, 8, 10, 11, 13, 17, 19, 23, 37, 61, 131,
-            4099 };
-
     /**
      * Test different array segment sizes, offsets, and number of negative
      * bytes.
      */
     public static void test_hasNegatives() throws Exception {
-        int len, off;
-        int ng;
-        boolean r;
-
-        for (ng = 0; ng < 57; ++ng) { // number of negatives in array segment
-            for (off = 0; off < 8; ++off) { // starting offset of array segment
-                for (int i = 0; i < sizes.length; ++i) { // array segment size
-                                                         // choice
-                    len = sizes[i];
-                    if (len + off > tBa.length)
-                        continue;
-                    initialize(off, len, ng);
-                    r = Helper.StringCodingHasNegatives(tBa, off, len);
-                    if (r ^ ((ng == 0) ? false : true)) {
-                        throw new Exception("Failed test hasNegatives " + "offset: " + off + " "
-                                + "length: " + len + " " + "return: " + r + " " + "negatives: "
-                                + ng);
-                    }
-                }
+        for (int off = 0; off < 16; off++) { // starting offset of array segment
+            // Test all array segment sizes 1-63
+            for (int len = 1; len < 64; len++) {
+                test_hasNegatives(off, len, 0);
+                test_hasNegatives(off, len, 1);
+                test_hasNegatives(off, len, RANDOM.nextInt(30) + 2);
             }
+            // Test a random selection of sizes between 64 and 4099, inclusive
+            for (int i = 0; i < 20; i++) {
+                int len = 64 + RANDOM.nextInt(4100 - 64);
+                test_hasNegatives(off, len, 0);
+                test_hasNegatives(off, len, 1);
+                test_hasNegatives(off, len, RANDOM.nextInt(len) + 2);
+            }
+        }
+    }
+
+    private static void test_hasNegatives(int off, int len, int maxNegatives) throws Exception {
+        assert (len + off < bytes.length);
+        initialize(off, len, maxNegatives);
+        boolean r = Helper.StringCodingHasNegatives(bytes, off, len);
+        if (r ^ ((maxNegatives == 0) ? false : true)) {
```

Comment:
This condition looks very obscure. Is it not equivalent to: `r ^ maxNegatives != 0`
Or even `r == (maxNegatives == 0)`?

Reply:
The `countPositives` intrinsics were adapted from the pre-existing `hasNegatives` intrinsic, and to ease porting over all platforms - including those we don't maintain - we specified `countPositives` so that if there's a negative value at index `idx` then it's acceptable for the intrinsics to return any value from `0` to `idx - 1`. This way intrinsics on all platforms could quickly be adapted over (return either `0` for `hasNegatives == true` or `length` for `hasNegatives == false`) while respective maintainers worked on porting over properly. 

This also allows the intrinsic to do a quick approximation: E.g. if a vectorized loop has checked 128 bytes but then detects a negative in the next block then the intrinsic can opt to return 128 rather than whatever the exact count. The surrounding java code will then do a scalar loop that is roughly as fast as an intrinsic tail routine would be, and with less complexity. I believe the `aarch64` intrinsic is implemented like this.